spring-messaging/src/main/java/org/springframework/messaging/simp/broker/DefaultSubscriptionRegistry.java

```diff
@@ -208,304 +214,250 @@ public void unregisterAllSubscriptions(String sessionId) {
+						result.add(sessionId, subscription.getId());
 					}
 				}
-				catch (Throwable ex) {
-					logger.debug("Failed to evaluate selector", ex);
-				}
 			}
 		});
+
 		return result;
 	}
 
-	@Override
-	public String toString() {
-		return "DefaultSubscriptionRegistry[" + this.destinationCache + ", " + this.subscriptionRegistry + "]";
+	private boolean evaluateExpression(@Nullable Expression expression, Message<?> message) {
+		boolean result = false;
+		try {
+			if (expression == null || Boolean.TRUE.equals(expression.getValue(messageEvalContext, message, Boolean.class))) {
+				result = true;
+			}
+		}
+		catch (SpelEvaluationException ex) {
+			if (logger.isDebugEnabled()) {
+				logger.debug("Failed to evaluate selector: " + ex.getMessage());
+			}
+		}
+		catch (Throwable ex) {
+			logger.debug("Failed to evaluate selector", ex);
+		}
+		return result;
 	}
 
-
 	/**
 	 * A cache for destinations previously resolved via
 	 * {@link DefaultSubscriptionRegistry#findSubscriptionsInternal(String, Message)}.
 	 */
-	private class DestinationCache {
+	private final class DestinationCache {
 
 		/** Map from destination to {@code <sessionId, subscriptionId>} for fast look-ups. */
-		private final Map<String, LinkedMultiValueMap<String, String>> accessCache =
+		private final Map<String, LinkedMultiValueMap<String, String>> destinationCache =
 				new ConcurrentHashMap<>(DEFAULT_CACHE_LIMIT);
 
-		/** Map from destination to {@code <sessionId, subscriptionId>} with locking. */
-		@SuppressWarnings("serial")
-		private final Map<String, LinkedMultiValueMap<String, String>> updateCache =
-				new LinkedHashMap<String, LinkedMultiValueMap<String, String>>(DEFAULT_CACHE_LIMIT, 0.75f, true) {
-					@Override
-					protected boolean removeEldestEntry(Map.Entry<String, LinkedMultiValueMap<String, String>> eldest) {
-						if (size() > getCacheLimit()) {
-							accessCache.remove(eldest.getKey());
-							return true;
-						}
-						else {
-							return false;
-						}
-					}
-				};
-
-
-		public LinkedMultiValueMap<String, String> getSubscriptions(String destination, Message<?> message) {
-			LinkedMultiValueMap<String, String> result = this.accessCache.get(destination);
-			if (result == null) {
-				synchronized (this.updateCache) {
-					result = new LinkedMultiValueMap<>();
-					for (SessionSubscriptionInfo info : subscriptionRegistry.getAllSubscriptions()) {
-						for (String destinationPattern : info.getDestinations()) {
-							if (getPathMatcher().match(destinationPattern, destination)) {
-								for (Subscription sub : info.getSubscriptions(destinationPattern)) {
-									result.add(info.sessionId, sub.getId());
-								}
-							}
-						}
-					}
-					if (!result.isEmpty()) {
-						this.updateCache.put(destination, result.deepCopy());
-						this.accessCache.put(destination, result);
+		private final Queue<String> cacheEvictionPolicy = new LinkedBlockingQueue<>();
```

Comment:
I did some perf testing of "LinkedBlockingQueue vs ConcurrentLinkedQueue" and I have no clear winner.
I think ConcurrentLinkedQueue.size() would cause lots of CPU cache miss while traversing through the whole linked list. That is something I don't know hot to test on local machine without some bigger test preparation. That is why I left LinkedBlockingQueue there, but I will change it to ConcurrentLinkedQueue if you like it more.

Reply:
finally, I used ConcurrentLinkedQueue and AtomicInteger to maintain the size.